dualstack support for GetPortAddresses

Billy99 opened the pull request:
Add "Pods" support for "pkg/ovn/pods.go:getPodAddresses() /
waitForPodAddresses() / pkg/util/net.go:GetPortAddresses() need to handle
multiple IPs".

Signed-off-by: Billy McFall <22157057+Billy99@users.noreply.github.com>

Billy99 commented:
This also includes (from #1142):
- pkg/ovn/pods.go:addLogicalPort()
  - updates for other pods.go and pod annotation changes
  - needs to pass both IPv4 and IPv6 addresses to lsp-set-addresses
  - needs to pass both IPv4 and IPv6 addresses to lsp-set-port-security

Billy99 commented:
@danwinship PTAL

danwinship commented on go-controller/pkg/ovn/pods.go:
```diff
@@ -250,12 +251,17 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) error {
+		if len(podCIDRs) == 1 {
+			addresses = fmt.Sprintf("%s %s", podMac, podCIDRs[0].IP)
+		} else {
+			addresses = fmt.Sprintf("%s %s %s", podMac, podCIDRs[0].IP, podCIDRs[1].IP)
```
as above, use `util.JoinIPs` rather than assuming the length

Billy99 replied:
Done

danwinship commented on go-controller/pkg/ovn/pods.go:
```diff
@@ -280,13 +286,20 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) error {
-		podCIDR = &net.IPNet{IP: podIP, Mask: nodeSubnet.Mask}
+		for _, podIP := range podIPs {
+			// DUAL-STACK FIXME: need multiple nodeSubnet.Mask
+			mask := nodeSubnet.Mask
```
#1283 merged yesterday and now `oc.waitForNodeLogicalSwitch` (which is where `nodeSubnet` came from) is nearly-dual-stack; you just need to fix it to actually return both subnets.

You'll need to match up the `nodeSubnets` with the `podIPs` correctly in terms of IPv4/IPv6; they're not guaranteed to be in the same order. A helper function like `gatewayForSubnet` in `pkg/ovn/gateway_init.go` might be useful.

Billy99 replied:
Was next on my list. I'll get right this.

danwinship commented on go-controller/pkg/ovn/pods.go:
```diff
@@ -305,10 +318,15 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) error {
+	if len(podCIDRs) == 1 {
+		addresses = fmt.Sprintf("%s %s", podMac, podCIDRs[0].IP)
+	} else {
+		addresses = fmt.Sprintf("%s %s %s", podMac, podCIDRs[0].IP, podCIDRs[1].IP)
```
as above...

Billy99 replied:
Done

danwinship commented on go-controller/pkg/util/net.go:
```diff
@@ -46,21 +46,30 @@ func GetPortAddresses(portName string) (net.HardwareAddr, net.IP, error) {
+	var ips []net.IP
+	var mac net.HardwareAddr
+	for i, addr := range addresses {
+		if i == 0 {
```
this is kind of weird... I think it would be better to keep the `addresess[0]` MAC case as it is, and then iterate over `range addresses[1:]` for the IPs

Billy99 replied:
Done.

Billy99 commented on go-controller/hybrid-overlay/pkg/controller/master.go:
```diff
@@ -122,20 +122,26 @@ func (m *MasterController) handleOverlayPort(node *kapi.Node, annotator kube.Ann
 		}
 		if portMAC == nil {
-			portMAC = util.IPAddrToHWAddr(portIP)
+			portMAC = util.IPAddrToHWAddr(portIPs[0])
```
Done

Billy99 commented on go-controller/hybrid-overlay/pkg/controller/master.go:
```diff
@@ -122,20 +122,26 @@ func (m *MasterController) handleOverlayPort(node *kapi.Node, annotator kube.Ann
+		if len(portIPs) == 1 {
+			addresses = fmt.Sprintf("%s %s", portMAC.String(), portIPs[0].String())
+		} else {
+			addresses = fmt.Sprintf("%s %s %s", portMAC.String(), portIPs[0].String(), portIPs[1].String())
```
Done. Much cleaner, thanks for the suggestion,

Billy99 commented on go-controller/hybrid-overlay/pkg/controller/node_linux.go:
```diff
@@ -449,21 +449,23 @@ func (n *NodeController) ensureHybridOverlayBridge() error {
+	for _, portIP := range portIPs {
+		portIPRaw := getIPAsHexString(portIP)
+		_, stderr, err = util.RunOVSOfctl("add-flow", extBridgeName,
+			fmt.Sprintf("table=0,priority=100,in_port=%s,arp,arp_tpa=%s,"+
```
Done

Billy99 commented on go-controller/pkg/ovn/pods.go:
```diff
@@ -229,14 +230,14 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) error {
 
 	var podMac net.HardwareAddr
-	var podCIDR *net.IPNet
+	var podCIDRs []*net.IPNet
```
Done

Billy99 commented on go-controller/pkg/ovn/pods.go:
```diff
@@ -229,14 +230,14 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) error {
 		podMac = annotation.MAC
-		// DUAL-STACK FIXME: handle multiple IPs
-		podCIDR = annotation.IPs[0]
+		podCIDRs = append(podCIDRs, annotation.IPs...)
```
Done.

Billy99 commented:
@danwinship PTAL
Thanks for the quick first review. When you have time, let me know if there is anything else you would like changed.

danwinship commented on go-controller/pkg/util/net.go:
```diff
@@ -140,13 +140,17 @@ func JoinIPs(ips []net.IP, sep string) string {
 
 // JoinIPNets joins the string forms of an array of *net.IPNet, as with strings.Join
-func JoinIPNets(ipnets []*net.IPNet, sep string) string {
+func JoinIPNets(ipnets []*net.IPNet, sep string, ipOnly bool) string {
```
boolean flags are tricky... I would have expected `true` to mean "show the subnet length" and `false` to mean "skip it".

Maybe just have a separate `JoinIPNetIPs()`?

Billy99 replied:
That's funny. I originally wrote it that way (separate function), then consolidated when self reviewing, thinking you would want less code. Will break back into separate function.

Billy99 commented:
@danwinship PTAL
Put `JoinIPNets()` back the way it was and added a new function for IP only.

danwinship commented on go-controller/pkg/ovn/pods.go:
```diff
@@ -337,15 +335,24 @@ func (oc *Controller) addLogicalPort(pod *kapi.Pod) error {
+
+			var gwIP net.IP
+			gwIfAddr := util.GetNodeGatewayIfAddr(nodeSubnet)
+			routes, gwIP, err = getRoutesGatewayIP(routes, pod, gwIfAddr)
```
It's weird that `routes` gets passed in, modified, and passed out, but `gwIPs` doesn't work the same way.

I'd suggested in #1142 that we could construct a partial `util.PodAnnotation` first, and then pass it to `getRoutesGatewayIP` to be modified (which would probably mean renaming it to `addRoutesGatewayIP` or something).

Billy99 replied:
Yes, I saw you note in #1142. I was trying to keep the amount of changes in each commit small and was going to follow up with that fix. I totally forgot to add a comment to code indicating that. Went ahead and added the suggested change now.

Done.

danwinship commented on go-controller/hybrid-overlay/pkg/controller/node_linux.go:
```diff
@@ -458,21 +459,26 @@ func (n *NodeController) ensureHybridOverlayBridge() error {
-		return fmt.Errorf("failed to set up hybrid overlay bridge ARP flow,"+
-			"stderr: %q, error: %v", stderr, err)
+	for _, portIP := range portIPs {
+		// DUAL-STACK FIXME: need IPv6 ARP support
```
just "need IPv6 support" (IPv6 doesn't use ARP, it uses somethint different)

Billy99 replied:
Done

trozet commented on go-controller/hybrid-overlay/pkg/controller/node_linux.go:
```diff
@@ -458,21 +459,26 @@ func (n *NodeController) ensureHybridOverlayBridge() error {
-	if err != nil {
-		return fmt.Errorf("failed to set up hybrid overlay bridge ARP flow,"+
-			"stderr: %q, error: %v", stderr, err)
+	for _, portIP := range portIPs {
```
going to need to hold on this patch and rebase after the external gw stuff goes in

Billy99 commented on go-controller/pkg/util/net.go:
```diff
@@ -168,3 +169,18 @@ func JoinIPNetsIPOnly(ipnets []*net.IPNet, sep string) string {
+
+// MatchIpFamily loops through the array of *net.IPNet and returns the
+// first entry in the list in the same IP Family, based on input flag isIPv6.
+func MatchIpFamily(isIPv6 bool, subnets []*net.IPNet) (*net.IPNet, error) {
```
Done

Billy99 commented:
@danwinship PTAL
Rebased after the External GW merge.

One question, in pod.go, moved `func addRoutesGatewayIP()` to be a member of the Controller interface (`func (oc *Controller) addRoutesGatewayIP()`) so the function could call `oc.getHybridOverlayExternalGwAnnotation()`. Do you prefer to pass in the `hybridOverlayExternalGW net.IP` and leave the call to `oc.getHybridOverlayExternalGwAnnotation()` in `addLogicalPort()` even though nothing else in the functions needs this value?

danwinship commented on go-controller/pkg/ovn/pods.go:
```diff
@@ -170,42 +169,67 @@ func getRoutesGatewayIP(pod *kapi.Pod, subnet *net.IPNet, hybridOverlayExternalG
+			// Add a route for each hybrid overlay subnet via the hybrid
+			// overlay port on the pod's logical switch.
+			second := util.NextIP(gatewayIP)
+			thirdIP := util.NextIP(second)
```
more bad existing code to fix.

    nextHop := util.GetNodeHybridOverlayIfAddr(nodeSubnet)

(and then `NextHop: nextHop.IP` below)

Actually, we shouldn't be checking `gatewayIP != nil` for this `if` block either. If there are hybrid overlay cluster subnets, then we _must_ route traffic for them through the hybrid overlay gateway, regardless of whether `otherDefaultRoute || hybridOverlayExternalGW != nil` was true above.

danwinship commented on go-controller/pkg/util/net.go:
```diff
@@ -150,3 +150,16 @@ func JoinIPNets(ipnets []*net.IPNet, sep string) string {
+
+// JoinIPNetsIPOnly joins the string forms of an array of *net.IPNet,
+// as with strings.Join, but does not include the IP mask.
+func JoinIPNetsIPOnly(ipnets []*net.IPNet, sep string) string {
```
maybe `JoinIPNetIPs`? It's a little shorter. But maybe it's less clear? Your call.

Billy99 commented:
/retest

Billy99 commented:
/retest

Billy99 commented:
/retest

danwinship commented:
lgtm. anyone else care to review before I merge it?
